### PR TITLE
Reactivate lubw dl

### PR DESCRIPTION
**Pull request by jdk24**
Draft as the API restricts usage to specific IPs. Will change after testing.

### CODE REVIEW

**Comment by jdk24** [comment]
Caveat for testing: Download only allowed from specific IPs. Not sure if you are included. 
Tested just now on new deployment:
Data table
Lookup table: It's actually the LUBW stations

File: lubw_download/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM alpine:latest
+# install dependencies
+# del preinstalled buybox sed with limited functionality
```
Comment (JakobMiksch) [inline lubw_download/Dockerfile:3]: do you mean `busybox` ?

File: lubw_download/resources/scripts/dl_lubw.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+#Set dirs and name parsed file after ingestion time
+OUTDIR=/data/temp
+OUTFILE=lubw_temp.xml
+PARSED_OUTDIR=/data/xml_data
+PARSED_OUTFILE=lubw_$(TZ=Europe/Berlin date +%Y%m%d%H%M).xml
+
+# Download xml file from server. Exit if curl throws failure (http response!=2xx) 
+
+if curl --fail -u $LUBW_USER:$LUBW_PASSWORD ftp://$LUBW_SERVER/Aktuelledaten.xml -o $OUTDIR/$OUTFILE;
```
Comment (JakobMiksch) [inline lubw_download/resources/scripts/dl_lubw.sh:61]: variables should ideally be in quotes. See [here](https://github.com/koalaman/shellcheck/wiki/SC2086)

you can use https://www.shellcheck.net/ to check the script. `shellcheck` is also availbable offline as cli program.

File: lubw_download/resources/scripts/dl_lubw.sh [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	echo "Datei $PARSED_OUTFILE heruntergeladen."
+else
+	echo "Error: Datei konnte nicht herunterladen werden."
+	exit 1
+fi
+
+
+# XML is not well formed for direct insert to PG. Need to replace newline with empty string.
+tr '\n' ' ' < $OUTDIR/$OUTFILE > $PARSED_OUTDIR/$PARSED_OUTFILE
+sleep 1
```
Comment (JakobMiksch) [inline lubw_download/resources/scripts/dl_lubw.sh:72]: why sleep?

File: lubw_download/resources/scripts/dl_lubw.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	elif [ "${!fileVar:-}" ]; then
+		val="$(< "${!fileVar}")"
+	fi
+	#echo "$var,$val"
+	export "$var"="$val"
+	unset "$fileVar" 
+}
+
+
+#Read docker secrets.
```
Comment (JakobMiksch) [inline lubw_download/resources/scripts/dl_lubw.sh:26]:
```suggestion
# Read docker secrets. 
```

File: lubw_download/resources/scripts/dl_lubw.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	LUBW_USER
+	LUBW_PASSWORD
+	LUBW_SERVER
+)
+
+for e in "${secrets[@]}"; do
+		file_env "$e"
+done
+
+#Ensure mandatory environment vars are set
```
Comment (JakobMiksch) [inline lubw_download/resources/scripts/dl_lubw.sh:38]:
```suggestion
# Ensure mandatory environment vars are set  
```

File: lubw_download/resources/scripts/dl_lubw.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+for e in "${envs[@]}"; do
+	if [ -z ${!e:-} ]; then
+		echo "error: $e is not set"
+		exit 1
+	fi
+done
+
+#Set dirs and name parsed file after ingestion time
```
Comment (JakobMiksch) [inline lubw_download/resources/scripts/dl_lubw.sh:53]:
```suggestion
# Set dirs and name parsed file after ingestion time
```

File: lubw_download/resources/scripts/dl_lubw.sh [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+		echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
+		exit 1
+	fi
+	local val="$def"
+	if [ "${!var:-}" ]; then
+		val="${!var}"
+	elif [ "${!fileVar:-}" ]; then
+		val="$(< "${!fileVar}")"
+	fi
+	#echo "$var,$val"
```
Comment (JakobMiksch) [inline lubw_download/resources/scripts/dl_lubw.sh:20]: remove unused line?

File: db/update-files.d/u12_alter_lubw_parser.sql [hunk 8]
```diff
@@ -0,0 +1,143 @@
+CREATE OR REPLACE FUNCTION station_data.lubw_parse()
```
Comment (JakobMiksch) [inline db/update-files.d/u12_alter_lubw_parser.sql:1]: some comments about this function would be useful to understand it later on

**Review by JakobMiksch — APPROVED** [review]
Great work! I added some (mostly cosmetic) remarks. The SQL function looks impressive :1st_place_medal: , but I am not familiar enough with it to properly review.
I have not tested the PR locally, but it should be fine to deploy, since it is a separate service that does not affect anything else.

**Review by chrismayer — APPROVED** [review]
LGTM, thnaks for this @jdk24 

Code looks good, could not test due to limited IP access.

File: lubw_download/resources/scripts/dl_lubw.sh (hunk 3 above)
Comment (jdk24) [reply lubw_download/resources/scripts/dl_lubw.sh:71]: This is just a precaution for `tr` reading from / outputting to a redirect, to properly finish parsing both files.
I encountered problems of reading from file streams where the operation finished before the end of file, which could also not be fixed by `wait `

Comment (JakobMiksch) [reply lubw_download/resources/scripts/dl_lubw.sh:71]: alright :+1:

**Review by JakobMiksch — APPROVED** [review]

File: db/update-files.d/u12_alter_lubw_parser.sql (hunk 8 above)
Comment (jdk24) [reply db/update-files.d/u12_alter_lubw_parser.sql]: Added comments on structure in [commit](https://github.com/meggsimum/sauber-sdi-docker/pull/120/commits/715cd123f1563ce15fed1187a3691b2cb0eebf30)

File: lubw_download/resources/scripts/dl_lubw.sh (hunk 7 above)
Comment (jdk24) [reply lubw_download/resources/scripts/dl_lubw.sh]: Removed commented line in [commit](https://github.com/meggsimum/sauber-sdi-docker/pull/120/commits/715cd123f1563ce15fed1187a3691b2cb0eebf30)

File: lubw_download/resources/scripts/dl_lubw.sh (hunk 6 above)
Comment (jdk24) [reply lubw_download/resources/scripts/dl_lubw.sh]: Applied in [commit](https://github.com/meggsimum/sauber-sdi-docker/pull/120/commits/715cd123f1563ce15fed1187a3691b2cb0eebf30)

File: lubw_download/resources/scripts/dl_lubw.sh (hunk 5 above)
Comment (jdk24) [reply lubw_download/resources/scripts/dl_lubw.sh]: Applied in [commit](https://github.com/meggsimum/sauber-sdi-docker/pull/120/commits/715cd123f1563ce15fed1187a3691b2cb0eebf30)

File: lubw_download/resources/scripts/dl_lubw.sh (hunk 4 above)
Comment (jdk24) [reply lubw_download/resources/scripts/dl_lubw.sh]: Thank you! Changes applied in bulk in new [commit](https://github.com/meggsimum/sauber-sdi-docker/pull/120/commits/715cd123f1563ce15fed1187a3691b2cb0eebf30)

**Pull request merged**